java-checks/src/main/java/org/sonar/java/closeresource/ExecutionState.java

```diff
@@ -0,0 +1,269 @@
+      Symbol symbol = entry.getKey();
+      CloseableOccurence currentOccurence = getCloseableOccurence(symbol);
+      CloseableOccurence occurenceToMerge = entry.getValue();
+      if (currentOccurence != null) {
+        currentOccurence.state = currentOccurence.state.merge(occurenceToMerge.state);
+        closeableOccurenceBySymbol.put(symbol, currentOccurence);
+      } else if (occurenceToMerge.state.isOpen()) {
+        insertIssue(occurenceToMerge.lastAssignment);
+      }
+    }
+    return this;
+  }
+
+  public ExecutionState overrideBy(ExecutionState currentES) {
+    for (Map.Entry<Symbol, CloseableOccurence> entry : currentES.closeableOccurenceBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      CloseableOccurence occurence = entry.getValue();
+      if (getCloseableOccurence(symbol) != null) {
+        markAs(symbol, occurence.state);
+      } else {
+        closeableOccurenceBySymbol.put(symbol, occurence);
+      }
+    }
+    return this;
+  }
+
+  public ExecutionState restoreParent() {
+    if (parent != null) {
+      insertIssues();
+      return parent.merge(this);
+    }
+    return this;
+  }
+
+  void insertIssues() {
+    for (Tree tree : getUnclosedClosables()) {
+      insertIssue(tree);
+    }
+  }
+
+  void insertIssue(Tree tree) {
+    Type type;
+    if (tree.is(Tree.Kind.VARIABLE)) {
+      type = ((VariableTree) tree).symbol().type();
+    } else {
+      type = ((IdentifierTree) tree).symbol().type();
+    }
+    check.addIssue(tree, "Close this \"" + type.name() + "\"");
+  }
+
+  void addCloseable(Symbol symbol, Tree lastAssignmentTree, @Nullable ExpressionTree assignmentExpression) {
+    CloseableOccurence newOccurence = new CloseableOccurence(lastAssignmentTree, getCloseableStateFromExpression(symbol, assignmentExpression));
+    CloseableOccurence knownOccurence = getCloseableOccurence(symbol);
+    if (knownOccurence != null) {
+      CloseableOccurence currentOccurence = closeableOccurenceBySymbol.get(symbol);
+      if (currentOccurence != null && currentOccurence.state.isOpen()) {
+        insertIssue(knownOccurence.lastAssignment);
+      }
+      if (!knownOccurence.state.isIgnored()) {
+        closeableOccurenceBySymbol.put(symbol, newOccurence);
+      }
+    } else {
+      closeableOccurenceBySymbol.put(symbol, newOccurence);
+    }
+  }
+
+  private State getCloseableStateFromExpression(Symbol symbol, @Nullable ExpressionTree expression) {
+    if (shouldBeIgnored(symbol, expression)) {
+      return State.IGNORED;
+    } else if (isNull(expression)) {
+      return State.NULL;
+    } else if (expression.is(Tree.Kind.NEW_CLASS)) {
+      if (usesIgnoredCloseableAsArgument(((NewClassTree) expression).arguments())) {
+        return State.IGNORED;
+      }
+      return State.OPEN;
+    }
+    // TODO SONARJAVA-1029 : Engine currently ignore closeables which are retrieved from method calls. Handle them as OPEN.
```

Comment:
:new: ![INFO](https://raw.githubusercontent.com/henryju/image-hosting/master/info.png)Complete the task associated to this TODO comment.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)